tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11144,6 +11144,51 @@ bool Sema::DiagnoseHLSLMethodCall(const CXXMethodDecl *MD, SourceLocation Loc) {
   return false;
 }
 
+void Sema::DiagnoseSVForLaunchType(const FunctionDecl *FD,
+                                   DXIL::NodeLaunchType LaunchTy) {
+  // Validate Compute Shader system value inputs per launch mode
+  for (ParmVarDecl *param : FD->parameters()) {
+    for (const hlsl::UnusualAnnotation *it : param->getUnusualAnnotations()) {
+      if (it->getKind() == hlsl::UnusualAnnotation::UA_SemanticDecl) {
+        const hlsl::SemanticDecl *sd = cast<hlsl::SemanticDecl>(it);
+        // if the node launch type is Thread, then there are no system values
+        // allowed
+        if (LaunchTy == DXIL::NodeLaunchType::Thread) {
+          if (sd->SemanticName.equals("SV_GroupThreadID") ||
+              sd->SemanticName.equals("SV_GroupIndex") ||
+              sd->SemanticName.equals("SV_GroupID") ||
+              sd->SemanticName.equals("SV_DispatchThreadID")) {
+            // emit diagnostic
+            unsigned DiagID = Diags.getCustomDiagID(
+                DiagnosticsEngine::Error,
+                "Invalid system value semantic '%0' for launchtype '%1'");
+            Diags.Report(param->getLocation(), DiagID)
+                << sd->SemanticName << "Thread";
+          }
+        }
+
+        // if the node launch type is Coalescing, then
+        // SV_GroupIndex and SV_GroupThreadID are allowed
+        else if (LaunchTy == DXIL::NodeLaunchType::Coalescing) {
+          if (sd->SemanticName.equals("SV_GroupID") ||
+              sd->SemanticName.equals("SV_DispatchThreadID")) {
```

Comment:
Similar to the above, should we be checking for the few that are invalid and expected or the two that are valid and erroring on all that aren't those?

Reply:
Excellent suggestion!